src/backends/gitlab/implementation.js

```diff
@@ -0,0 +1,148 @@
+import { CURSOR_COMPATIBILITY_SYMBOL } from 'ValueObjects/Cursor';
+import { EDITORIAL_WORKFLOW } from "Constants/publishModes";
+
+const MAX_CONCURRENT_DOWNLOADS = 10;
+
+export default class GitLab {
+  constructor(config, proxied = false) {
+    this.config = config;
+
+    if (config.getIn(["publish_mode"]) === EDITORIAL_WORKFLOW) {
+      throw new Error("The GitLab backend does not support the Editorial Workflow.")
+    }
+
+    if (!proxied && config.getIn(["backend", "repo"]) == null) {
+      throw new Error("The GitLab backend needs a \"repo\" in the backend configuration.");
+    }
+
+    this.repo = config.getIn(["backend", "repo"], "");
+    this.branch = config.getIn(["backend", "branch"], "master");
+    this.api_root = config.getIn(["backend", "api_root"], "https://gitlab.com/api/v4");
+    this.token = '';
+  }
+
+  authComponent() {
+    return AuthenticationPage;
+  }
+
+  restoreUser(user) {
+    return this.authenticate(user);
+  }
+
+  authenticate(state) {
+    this.token = state.token;
+    this.api = new API({ token: this.token, branch: this.branch, repo: this.repo, api_root: this.api_root });
+    return this.api.user().then(user =>
+      this.api.hasWriteAccess(user).then((isCollab) => {
+        // Unauthorized user
+        if (!isCollab) throw new Error("Your GitLab user account does not have access to this repo.");
+        // Authorized user
+        return Object.assign({}, user, { token: state.token });
+      })
+    );
+  }
+
+  logout() {
+    this.token = null;
+    return;
+  }
+
+  getToken() {
+    return Promise.resolve(this.token);
+  }
+
+  entriesByFolder(collection, extension) {
+    return this.api.listFiles(collection.get("folder"))
+    .then(({ files, cursor }) =>
+      this.fetchFiles(files.filter(file => fileExtension(file.name) === extension))
+      .then(fetchedFiles => {
+        const returnedFiles = fetchedFiles;
+        returnedFiles[CURSOR_COMPATIBILITY_SYMBOL] = cursor;
+        return returnedFiles;
+      })
+    );
+  }
+
+  allEntriesByFolder(collection, extension) {
+    return this.api.listAllFiles(collection.get("folder"))
+    .then(files => this.fetchFiles(files.filter(file => fileExtension(file.name) === extension)));
+  }
+
+  entriesByFiles(collection) {
+    const files = collection.get("files").map(collectionFile => ({
+      path: collectionFile.get("file"),
+      label: collectionFile.get("label"),
+    }));
+    return this.fetchFiles(files).then(fetchedFiles => {
+      const returnedFiles = fetchedFiles;
+      return returnedFiles;
+    });
+  }
+
+  fetchFiles = (files) => {
+    const sem = semaphore(MAX_CONCURRENT_DOWNLOADS);
+    const promises = [];
+    files.forEach((file) => {
+      promises.push(new Promise((resolve, reject) => (
+        sem.take(() => this.api.readFile(file.path, file.id).then((data) => {
+          resolve({ file, data });
+          sem.leave();
+        }).catch((err) => {
+          sem.leave();
+          reject(err);
```

Comment:
I think we need to pull in the code from the GitHub `fetchFiles` instead of rejecting here. That was changed to fix the problem where a directory is empty if a single file is missing/broken.

Reply:
https://github.com/netlify/netlify-cms/pull/1093 will need ported into here.